### PR TITLE
Fix `loc` for Series with a MultiIndex

### LINKED ISSUE #7221: [BUG] `loc` fails on a Series with a MultiIndex
Using a tuple to index into a Series with a MultiIndex fails:

```python
In [9]: a = cudf.DataFrame({'a': [1, 2], 'b': [2, 3], 'c': ['a', 'b']}).set_index(['a', 'b'])['c']

In [10]: a
Out[10]:
a  b
1  2    a
2  3    b
Name: c, dtype: object

In [11]: a.loc[(1, 2)]  # AttributeError

In [12]: a.to_pandas().loc[(1, 2)]  # this works OK
Out[12]: 'a'
```

### PR DESCRIPTION
**Pull request by shwina**
Fixes #7221 and adds improvements to `loc` with a MultiIndex.

* Previously, `loc` on a `Series` with a `MultiIndex` would fail. For example:

```python
In [7]: sr
Out[7]:
n_workers  type
1          fit        1
2          load       2
3          predict    3
Name: x, dtype: int64

In [8]: sr.loc[(1, "fit")]  # KeyError
````

* Previously, `loc` on a `DataFrame` with a `MultiIndex` would fail when a slice without `start` or `end` was used. For example:

```python
In [3]: df
Out[3]:
                   x
n_workers type
1         fit      1
2         load     2
3         predict  3

In [4]: df.loc[:(2, "load")]  # TypeError
```

Both the above issues have been addressed and tests added.

### CODE REVIEW

File: python/cudf/cudf/core/indexing.py [hunk 1]
```diff
@@ -113,7 +116,19 @@ class _SeriesLocIndexer(object):
...
+    def __getitem__(self, arg: Any) -> Union[ScalarLike, DataFrameOrSeries]:
+        if isinstance(arg, (cudf.MultiIndex, pd.MultiIndex)):
+            if isinstance(arg, pd.MultiIndex):
+                arg = cudf.from_pandas(arg)
+        elif isinstance(self._sr.index, cudf.MultiIndex):
+            result = self._sr.index._get_row_major(self._sr, arg)
+            if (
+                isinstance(arg, tuple)
+                and len(arg) == self._sr._index.nlevels
+                and not any((isinstance(x, slice) for x in arg))
```
Comment (kkraus14) [inline python/cudf/cudf/core/indexing.py:128]: What is `arg` in this case? Could it be a GPU backed object?

Comment (shwina) [reply python/cudf/cudf/core/indexing.py:128]: A tuple (see 2 lines above) for indexing into the levels of a MultiIndex. Something like `s.loc[('a', 'b', slice(None), 'd')]`

File: python/cudf/cudf/core/indexing.py [hunk 2]
```diff
@@ -113,7 +116,19 @@ class _SeriesLocIndexer(object):
     def __init__(self, sr):
         self._sr = sr
 
-    def __getitem__(self, arg):
+    def __getitem__(self, arg: Any) -> Union[ScalarLike, DataFrameOrSeries]:
+        if isinstance(arg, (cudf.MultiIndex, pd.MultiIndex)):
+            if isinstance(arg, pd.MultiIndex):
+                arg = cudf.from_pandas(arg)
```
Comment (isVoid) [inline python/cudf/cudf/core/indexing.py:122]: Does this look like necessary branching?
```suggestion
        if isinstance(arg, pd.MultiIndex):
                arg = cudf.from_pandas(arg)
```

**Review by isVoid — COMMENTED** [review]
Perhaps PR description can provide a touch more detail on the improvement on `MultiIndex`?

File: python/cudf/cudf/core/indexing.py [hunk 3]
```diff
@@ -1,10 +1,13 @@
 # Copyright (c) 2020, NVIDIA CORPORATION.
+from typing import Any, Union
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/core/indexing.py:2]: copyright

File: python/cudf/cudf/tests/test_indexing.py [hunk 4]
```diff
@@ -1343,3 +1343,21 @@ def test_loc_zero_dim_array():
...
     assert_eq(psr[np.array([0])[0]], gsr[np.array([0])[0]])
+
+
+@pytest.mark.parametrize(
+    "arg",
+    [
+        slice(None),
+        slice((1, 2), None),
+        slice(None, (1, 2)),
+        (1, 1),
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/tests/test_indexing.py:1354]: Should we add negative test for tuple size > lvl in multi-indexing

File: python/cudf/cudf/core/indexing.py (hunk 2 above)
Comment (shwina) [reply python/cudf/cudf/core/indexing.py:122]: oops - nice catch!

**Review by kkraus14 — APPROVED** [review]
LGTM but should address @rgsl888prabhu's comments

**Review by isVoid — APPROVED** [review]

**Review by galipremsagar — APPROVED** [review]

**Review by rgsl888prabhu — APPROVED** [review]

File: python/cudf/cudf/core/indexing.py (hunk 3 above)
Comment (shwina) [reply python/cudf/cudf/core/indexing.py:3]: Fixed

File: python/cudf/cudf/tests/test_indexing.py (hunk 4 above)
Comment (shwina) [reply python/cudf/cudf/tests/test_indexing.py:1354]: Added better validation of the tuple and some tests

**Comment by kkraus14** [comment]
@gpucibot merge

**Comment by shwina** [comment]
rerun tests

**Comment by kkraus14** [comment]
@gpucibot merge

**Pull request merged**